nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1270,6 +1301,113 @@ mod test {
+        builder.policy_mut().target_nexus_zone_count = 1;
+
+        // Find whatever pool NTP was using
+        let pool_to_expunge = blueprint1
+            .blueprint_zones
+            .iter()
+            .find_map(|(_, zones_config)| {
+                for zone_config in &zones_config.zones {
+                    if zone_config.zone_type.is_ntp() {
+                        return zone_config.filesystem_pool.clone();
+                    }
+                }
+                None
+            })
+            .expect("No NTP zone pool?");
+
+        // This is mostly for test stability across "example system" changes:
+        // Find how many other zones are using this same zone.
+        let zones_using_zpool = blueprint1.blueprint_zones.iter().fold(
+            0,
+            |acc, (_, zones_config)| {
+                let mut zones_using_zpool = 0;
+                for zone_config in &zones_config.zones {
+                    if let Some(pool) = &zone_config.filesystem_pool {
+                        if pool == &pool_to_expunge {
+                            zones_using_zpool += 1;
+                            continue;
+                        }
+                    }
+                    if let Some(pool) = zone_config.zone_type.durable_zpool() {
+                        if pool == &pool_to_expunge {
+                            zones_using_zpool += 1;
+                            continue;
+                        }
+                    }
+                }
+                acc + zones_using_zpool
+            },
+        );
+        assert!(
+            zones_using_zpool > 0,
+            "We should be expunging at least one zone using this zpool"
+        );
+
+        // For that pool, find the physical disk behind it, and mark it
+        // expunged.
+        let (_, sled_details) = builder.sleds_mut().iter_mut().next().unwrap();
+        let disk = sled_details
+            .resources
+            .zpools
+            .get_mut(&pool_to_expunge.id())
+            .unwrap();
+        disk.policy = PhysicalDiskPolicy::Expunged;
+
+        let input = builder.build();
+
+        let blueprint2 = Planner::new_based_on(
+            logctx.log.clone(),
+            &blueprint1,
+            &input,
+            "test: expunge a disk with a zone on top",
+            &collection,
+        )
+        .expect("failed to create planner")
+        .with_rng_seed((TEST_NAME, "bp2"))
+        .plan()
+        .expect("failed to plan");
+
+        let diff = blueprint2.diff_since_blueprint(&blueprint1);
+        println!("1 -> 2 (expunge a disk):\n{}", diff.display());
+        assert_eq!(diff.sleds_added.len(), 0);
+        assert_eq!(diff.sleds_removed.len(), 0);
+        assert_eq!(diff.sleds_modified.len(), 1);
+
+        // We should be removing all zones using this zpool
+        assert_eq!(diff.zones.added.len(), 0);
+        assert_eq!(diff.zones.removed.len(), 0);
+        assert_eq!(diff.zones.modified.len(), 1);
```

Comment:
Should this instead be `assert_eq!(diff.zones.modified.len(), zones_using_zpool)`?

Reply:
I don't think so! I made this same mistake when originally writing the PR, and thought there was a bug with the blueprint diff process until I looked at the types.

`diff.zones.modified` is a BTreeMap of "Sled UUID" -> `BpDiffZonesModified`. If many zones are modified, but they're all on one sled, only a single entry should exist here.

(I am doing the check you suggest in the lines that follow, where I turn `diff.zones.modified` into an iterator, access the first sled, and then check that it has `zones_using_zpool` modified zones)

Reply:
Ha, awesome. I already forgot about that code!